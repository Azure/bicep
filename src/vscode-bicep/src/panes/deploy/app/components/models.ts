// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
export interface TemplateMetadata {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  template: any;
  parameters: ParamDefinition[];
}

export interface ParamDefinition {
  type: string;
  name: string;
  defaultValue?: string;
}

export interface ParamData {
  useDefault: boolean;
  value: string;
}
