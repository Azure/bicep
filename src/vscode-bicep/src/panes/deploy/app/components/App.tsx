// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { FC, useState } from "react";
import { VSCodeButton, VSCodeProgressRing } from "@vscode/webview-ui-toolkit/react";
import "./index.css";
import { ParamData } from "../../models";
import { useMessageHandler } from "./hooks/useMessageHandler";
import { WhatIfChangesView } from "./sections/WhatIfChangesView";
import { DeploymentOperationsView } from "./sections/DeploymentOperationsView";
import { DeploymentOutputsView } from "./sections/DeploymentOutputsView";
import { ResultsView } from "./sections/ResultsView";
import { ParametersInputView } from "./sections/ParametersInputView";
import { useAzure } from "./hooks/useAzure";
import { DeploymentScopeInputView } from "./sections/DeploymentScopeInputView";
import { FormSection } from "./sections/FormSection";

export const App: FC = () => {
  const [errorMessage, setErrorMessage] = useState<string>();
  const messages = useMessageHandler({ setErrorMessage });
  const azure = useAzure({
    scope: messages.scope,
    acquireAccessToken: messages.acquireAccessToken,
    templateMetadata: messages.templateMetadata,
    parametersMetadata: messages.paramsMetadata,
    setErrorMessage
  });

  function setParamValue(key: string, data: ParamData) {
    const parameters = Object.assign({}, messages.paramsMetadata.parameters, { [key]: data });
    messages.setParamsMetadata({ ...messages.paramsMetadata, parameters });
  }

  function handleEnableParamEditing() {
    messages.setParamsMetadata({ ...messages.paramsMetadata, sourceFilePath: undefined });
  }

  const azureDisabled = !messages.scope || !messages.templateMetadata || azure.running;

  async function handleDeployClick() {
    messages.publishTelemetry('deployPane/deploy', {});
    await azure.deploy();
  }

  async function handleValidateClick() {
    messages.publishTelemetry('deployPane/validate', {});
    await azure.validate();
  }

  async function handleWhatIfClick() {
    messages.publishTelemetry('deployPane/whatIf', {});
    await azure.whatIf();
  }

  return (
    <main id="webview-body">
      <FormSection title="Experimental Warning">
        <div className="alert-error">
          <span className="codicon codicon-beaker" />
          The Bicep Deployment Pane is an experimental feature.
          <br/>
          Documentation is available <a href="https://github.com/Azure/bicep/blob/main/docs/deploy-ui.md">here</a>. Please raise issues or feature requests <a href={createNewIssueUrl()}>here</a>.
        </div>
      </FormSection>
      <DeploymentScopeInputView
        scope={messages.scope}
        onPickScope={messages.pickScope} />

      <ParametersInputView
        parameters={messages.paramsMetadata}
        template={messages.templateMetadata}
        disabled={azure.running}
        onValueChange={setParamValue}
        onEnableEditing={handleEnableParamEditing}
        onPickParametersFile={messages.pickParamsFile} />

      <FormSection title="Actions">
        {errorMessage && <div className="alert-error">
          <span className="codicon codicon-error" />
          {errorMessage}
        </div>}
        <div className="controls">
          <VSCodeButton onClick={handleDeployClick} disabled={azureDisabled}>Deploy</VSCodeButton>
          <VSCodeButton onClick={handleValidateClick} disabled={azureDisabled}>Validate</VSCodeButton>
          <VSCodeButton onClick={handleWhatIfClick} disabled={azureDisabled}>What-If</VSCodeButton>
        </div>
        {azure.running && <VSCodeProgressRing></VSCodeProgressRing>}
      </FormSection>

      {messages.scope && <>
        <ResultsView result={azure.result} />
        <DeploymentOperationsView scope={messages.scope} operations={azure.operations} />
        <DeploymentOutputsView outputs={azure.outputs} />
        <WhatIfChangesView changes={azure.whatIfChanges} />
      </>}
    </main>
  );
};

function createNewIssueUrl() {
  const title = 'Deployment Pane: <description>'
  const labels = ['story: deploy pane']
  const body = `

`;

  return `https://github.com/Azure/bicep/issues/new?title=${encodeURIComponent(title)}&body=${encodeURIComponent(body)}&labels=${encodeURIComponent(labels.join(','))}`;
}