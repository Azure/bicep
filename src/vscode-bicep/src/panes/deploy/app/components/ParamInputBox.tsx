import { VSCodeCheckbox, VSCodeTextField } from "@vscode/webview-ui-toolkit/react";
import { FC, FormEvent } from "react";
import { ParamData, ParamDefinition } from "./models";

interface ParamInputBoxProps {
  definition: ParamDefinition;
  data: ParamData;
  disabled: boolean;
  onChangeData: (data: ParamData) => void;
}

export const ParamInputBox : FC<ParamInputBoxProps> = (props) => {
  const { definition, data, disabled, onChangeData } = props;
  function handleValueInput(event: FormEvent<HTMLInputElement>) {
    onChangeData({
      ...data,
      value: event.currentTarget.value,
    });
  }
  function handleDefaultValueChange() {
    onChangeData({
      value: definition.defaultValue ?? '',
      useDefault: !data.useDefault,
    });
  }

  const { name, defaultValue } = definition;
  const { value, useDefault } = data;

  return (
    <span className="input-row">
      <VSCodeTextField
        value={value ?? ''}
        onInput={e => handleValueInput(e as FormEvent<HTMLInputElement>)}
        disabled={disabled || useDefault}>
        {name}
      </VSCodeTextField>
      {(defaultValue !== undefined) ? (
        <VSCodeCheckbox
          checked={useDefault}
          onChange={handleDefaultValueChange}
          disabled={disabled}>
          Use Default?
        </VSCodeCheckbox>
      ) : null}
    </span>
  );
}