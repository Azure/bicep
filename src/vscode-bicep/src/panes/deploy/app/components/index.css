#webview-body {
  display: flex;
  flex-direction: column;
}

.form-section {
  display: grid;
  grid-auto-flow: row;
  grid-row-gap: 1rem;
  overflow-x: scroll;
}

.input-row {
  display: grid;
  grid-auto-flow: row;
}

@media screen and (min-width: 1024px) {
  #webview-body {
    padding: 0 8rem;
    max-width: 1024px;
    margin: 0 auto;
  }
}

vscode-button {
  width: fit-content;
  margin-right: 0.5rem;
}