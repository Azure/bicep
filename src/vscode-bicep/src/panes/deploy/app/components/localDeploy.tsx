// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { VSCodeDataGrid, VSCodeDataGridCell, VSCodeDataGridRow } from "@vscode/webview-ui-toolkit/react";
import { FC } from "react";
import { LocalDeploymentOperationContent, LocalDeployResponse } from "../../../../language";
import { FormSection } from "./sections/FormSection";
import { getPreformattedJson } from "./utils";

export const LocalDeployResult: FC<{ result: LocalDeployResponse }> = ({ result }) => {
  const error = result.deployment.error;
  return (
    <FormSection title="Result">
      <p>{result.deployment.provisioningState}</p>
      {error && (
        <VSCodeDataGrid gridTemplateColumns="max-content auto">
          {error && (
            <VSCodeDataGridRow key={0}>
              <VSCodeDataGridCell gridColumn="1">Code</VSCodeDataGridCell>
              <VSCodeDataGridCell gridColumn="2">{error.code}</VSCodeDataGridCell>
            </VSCodeDataGridRow>
          )}
          {error.message && (
            <VSCodeDataGridRow key={1}>
              <VSCodeDataGridCell gridColumn="1">Message</VSCodeDataGridCell>
              <VSCodeDataGridCell gridColumn="2">{error.message}</VSCodeDataGridCell>
            </VSCodeDataGridRow>
          )}
        </VSCodeDataGrid>
      )}
    </FormSection>
  );
};

export const LocalDeployOperations: FC<{ result: LocalDeployResponse }> = ({ result }) => {
  if (!result.operations) {
    return null;
  }

  return (
    <FormSection title="Operations">
      <VSCodeDataGrid>
        <VSCodeDataGridRow rowType="header">
          <VSCodeDataGridCell gridColumn="1" cellType="columnheader">
            Resource Name
          </VSCodeDataGridCell>
          <VSCodeDataGridCell gridColumn="2" cellType="columnheader">
            State
          </VSCodeDataGridCell>
          <VSCodeDataGridCell gridColumn="3" cellType="columnheader">
            Error
          </VSCodeDataGridCell>
        </VSCodeDataGridRow>
        {result.operations.map((operation) => (
          <VSCodeDataGridRow
            key={operation.resourceName}
            style={isFailed(operation) ? { background: "rgba(255, 72, 45, 0.3)" } : {}}
          >
            <VSCodeDataGridCell gridColumn="1">{operation.resourceName}</VSCodeDataGridCell>
            <VSCodeDataGridCell gridColumn="2">{operation.provisioningState}</VSCodeDataGridCell>
            <VSCodeDataGridCell gridColumn="3">
              {operation.error ? getPreformattedJson(operation.error) : ""}
            </VSCodeDataGridCell>
          </VSCodeDataGridRow>
        ))}
      </VSCodeDataGrid>
    </FormSection>
  );
};

export const LocalDeployOutputs: FC<{ result: LocalDeployResponse }> = ({ result }) => {
  if (!result.deployment.outputs) {
    return null;
  }

  return (
    <FormSection title="Outputs">
      <VSCodeDataGrid>
        <VSCodeDataGridRow rowType="header">
          <VSCodeDataGridCell gridColumn="1" cellType="columnheader">
            Name
          </VSCodeDataGridCell>
          <VSCodeDataGridCell gridColumn="2" cellType="columnheader">
            Value
          </VSCodeDataGridCell>
        </VSCodeDataGridRow>
        {Object.keys(result.deployment.outputs).map((name) => (
          <VSCodeDataGridRow key={name}>
            <VSCodeDataGridCell gridColumn="1">{name}</VSCodeDataGridCell>
            <VSCodeDataGridCell gridColumn="2">
              {getPreformattedJson(result.deployment.outputs[name])}
            </VSCodeDataGridCell>
          </VSCodeDataGridRow>
        ))}
      </VSCodeDataGrid>
    </FormSection>
  );
};

function isFailed(operation: LocalDeploymentOperationContent) {
  return operation.provisioningState.toLowerCase() === "failed";
}
