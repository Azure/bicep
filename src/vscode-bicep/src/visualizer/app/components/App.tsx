// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
// import cytoscape from "cytoscape";
import { useEffect, useState, VFC } from "react";
import { ElementDefinition } from "cytoscape";
import { DefaultTheme, ThemeProvider } from "styled-components";

import { StatusBar } from "./StatusBar";
import {
  Graph,
  createContainerNodeBackgroundUri,
  createChildlessNodeBackgroundUri,
} from "./Graph";

import { DeploymentGraphMessage, Message, READY_MESSAGE } from "../../messages";
import { DeploymentGraph } from "../../../language";
import { darkTheme, lightTheme, highContrastTheme } from "../themes";
import { vscode } from "../vscode";

async function mapToElements(
  graph: DeploymentGraphMessage["deploymentGraph"],
  theme: DefaultTheme
): Promise<ElementDefinition[]> {
  if (!graph) {
    return [];
  }

  const nodes = await Promise.all(
    graph.nodes.map(async (node) => {
      const idSegments = node.id.split("::");
      const symbol = idSegments.pop() as string;
      const parent = idSegments.length > 0 ? idSegments.join("::") : undefined;

      return {
        data: {
          id: node.id,
          parent,
          hasError: node.hasError,
          filePath: node.filePath,
          range: node.range,
          backgroundDataUri: node.hasChildren
            ? createContainerNodeBackgroundUri(symbol, node.isCollection, theme)
            : await createChildlessNodeBackgroundUri(
                symbol,
                node.type,
                node.isCollection,
                theme
              ),
        },
      };
    })
  );

  const edges = graph.edges.map(({ sourceId, targetId }) => ({
    data: {
      id: `${sourceId}>${targetId}`,
      source: sourceId,
      target: targetId,
    },
  }));

  return [...nodes, ...edges];
}

export const App: VFC = () => {
  const [elements, setElements] = useState<ElementDefinition[]>([]);
  const [graph, setGraph] = useState<DeploymentGraph | null>(null);
  const [theme, setTheme] = useState<DefaultTheme>(darkTheme);

  const handleMessageEvent = (e: MessageEvent<Message>) => {
    const message = e.data;
    if (message.kind === "DEPLOYMENT_GRAPH") {
      vscode.setState(message.documentPath);
      setGraph(message.deploymentGraph);
    }
  };

  const applyTheme = (bodyClassName: string) => {
    switch (bodyClassName) {
      case "vscode-dark":
        setTheme(darkTheme);
        break;
      case "vscode-light":
        setTheme(lightTheme);
        break;
      case "vscode-high-contrast":
        setTheme(highContrastTheme);
        break;
    }
  };

  useEffect(() => {
    mapToElements(graph, theme).then(setElements);
  }, [graph, theme]);

  useEffect(() => {
    window.addEventListener("message", handleMessageEvent);
    vscode.postMessage(READY_MESSAGE);
    return () => window.removeEventListener("message", handleMessageEvent);
  }, []);

  useEffect(() => {
    applyTheme(document.body.className);

    const observer = new MutationObserver((mutationRecords) =>
      mutationRecords.forEach((mutationRecord) =>
        applyTheme((mutationRecord.target as HTMLElement).className)
      )
    );

    observer.observe(document.body, {
      attributes: true,
      attributeFilter: ["class"],
    });

    return () => observer.disconnect();
  }, []);

  return (
    <ThemeProvider theme={theme}>
      <Graph theme={theme} elements={elements} />
      <StatusBar
        errorCount={graph?.errorCount ?? 0}
        hasNodes={elements.length > 0}
      />
    </ThemeProvider>
  );
};
