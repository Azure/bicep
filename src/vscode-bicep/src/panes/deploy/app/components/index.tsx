// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import ReactDOM from "react-dom";

import { App } from "./App";

ReactDOM.render(
  <App />,
  document.getElementById("root")
);
