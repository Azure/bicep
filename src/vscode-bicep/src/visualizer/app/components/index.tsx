// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import ReactDOM from "react-dom";
import { createGlobalStyle } from "styled-components";

import { App } from "./App";

const GlobalStyle = createGlobalStyle`
  body {
    height: 100vh;
    margin: 0;
    padding: 0;
    display: flex;
    overflow: hidden;
  }

  #root {
    flex: 1 1 auto;
    overflow: hidden;
  }
`;

ReactDOM.render(
  <>
    <GlobalStyle />
    <App />
  </>,
  document.getElementById("root")
);
