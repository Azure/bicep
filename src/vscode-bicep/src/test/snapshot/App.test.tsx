// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import renderer from "react-test-renderer";
import "jest-styled-components";

import { App } from "../../visualizer/app/components/App";

describe("component App", () => {
  it("should render", () => {
    const tree = renderer.create(<App />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
