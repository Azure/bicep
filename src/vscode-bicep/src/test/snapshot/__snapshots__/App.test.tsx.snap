// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component App should render 1`] = `
Array [
  .c0 {
  position: absolute;
  left: 0px;
  top: 0px;
  bottom: 0px;
  right: 0px;
  overflow: hidden;
  background-color: #111111;
  background-image: radial-gradient(circle at 1px 1px,#3f3f3f 1px,transparent 0);
  background-size: 24px 24px;
  background-position: 12px 12px;
}

<div
    className="c0"
  />,
  .c0 {
  position: absolute;
  height: 32px;
  left: 20px;
  bottom: 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  width: 8px;
  height: 8px;
  background-color: Green;
  border-radius: 50%;
  color: white;
  margin-top: 2px;
  margin-right: 8px;
}

<div
    className="c0"
  >
    <div
      className="c1"
    />
    <div>
      There is no resources or modules in the file. Nothing to render.
    </div>
  </div>,
]
`;
